### PR TITLE
Added SET and LIST cardinality support

### LINKED ISSUE #127: Cardinality Support?
[This Cardinality](http://tinkerpop.apache.org/javadocs/3.2.1/core/index.html?org/apache/tinkerpop/gremlin/structure/VertexProperty.Cardinality.html) to be clear

Right now it there doesn't seem to be deriving elements that are `Vec<T>` or `HashSet<T>` on structs as in stuff like `Vec<String>` or a `HashSet<String>`. 

I was able to prove out something that appears to work with these:

```

fn for_list<T>(glist: List) -> GremlinResult<Vec<T>>
where
    T: std::convert::TryFrom<GValue, Error = GremlinError>,
{
    glist
        .take()
        .into_iter()
        .map(|x| x.try_into())
        .collect::<GremlinResult<Vec<T>>>()
}

fn for_list_to_set<T>(glist: List) -> GremlinResult<HashSet<T>>
where
    T: std::convert::TryFrom<GValue, Error = GremlinError> + Hash + Eq,
{
    glist
        .take()
        .into_iter()
        .map(|x| x.try_into())
        .collect::<GremlinResult<HashSet<T>>>()
}

fn for_set<T>(gset: Set) -> GremlinResult<HashSet<T>>
where
    T: std::convert::TryFrom<GValue, Error = GremlinError> + Hash + Eq,
{
    gset.take()
        .into_iter()
        .map(|x| x.try_into())
        .collect::<GremlinResult<HashSet<T>>>()
}


macro_rules! impl_try_from_set {
    ($t:ty) => {
        impl std::convert::TryFrom<GValue> for HashSet<$t> {
            type Error = crate::GremlinError;

            fn try_from(value: GValue) -> GremlinResult<Self> {
                match value {
                    GValue::List(s) => for_list_to_set(s),
                    GValue::Set(s) => for_set(s),
                    _ => Err(GremlinError::Cast(format!(
                        "Cannot cast {:?} to Set",
                        value
                    ))),
                }
            }
        }
    };
}

impl_try_from_set!(String);
impl_try_from_set!(i32);
impl_try_from_set!(i64);

macro_rules! impl_try_from_list {
    ($t:ty) => {
        impl std::convert::TryFrom<GValue> for Vec<$t> {
            type Error = crate::GremlinError;

            fn try_from(value: GValue) -> GremlinResult<Self> {
                match value {
                    GValue::List(s) => for_list(s),
                    _ => Err(GremlinError::Cast(format!(
                        "Cannot cast {:?} to Vec",
                        value
                    ))),
                }
            }
        }
    };
}

impl_try_from_list!(String);
impl_try_from_list!(i32);
impl_try_from_list!(i64);
```

I could do another PR if there's interest in incorporating these. FWIW I was somewhat confused though by the logic here:

https://github.com/wolf4ood/gremlin-rs/blob/2e18b3cb09217a1c7f9645f98ce98a90a69acfab/gremlin-client/src/structure/value.rs#L385

Seems like we shouldn't be attempting to parse List properties there? Maybe the PR should also remove that logic?

### PR DESCRIPTION
**Pull request by criminosis**
Closes #127

### CODE REVIEW

File: gremlin-client/src/structure/value.rs [hunk 1]
```diff
@@ -408,3 +409,121 @@ macro_rules! impl_try_from_option {
...
+                        value
+                    ))),
+                }
+            }
+        }
+    };
+}
+
+impl_try_from_list!(String);
+// impl_try_from_list!(i32);
```
Comment (criminosis) [inline gremlin-client/src/structure/value.rs:523]: Still WIP for these types, going to add them to the tests next, but wanted to show the work so far for feedback

**Comment by criminosis** [comment]
@wolf4ood ready for your review at your convenience

File: gremlin-client/src/conversion.rs [hunk 2]
```diff
@@ -117,6 +119,7 @@ impl_from_gvalue!(f32, GValue::Float);
 impl_from_gvalue!(f64, GValue::Double);
 impl_from_gvalue!(i32, GValue::Int32);
 impl_from_gvalue!(i64, GValue::Int64);
+impl_from_gvalue!(bool, GValue::Bool);
```
Comment (criminosis) [inline gremlin-client/src/conversion.rs:122]: Not strictly related to cardinality but its omission was causing missing trait compiler errors, hope it's okay I added it.

File: gremlin-client/src/structure/value.rs [hunk 3]
```diff
@@ -333,6 +334,36 @@ impl std::convert::TryFrom<GValue> for i64 {
     }
 }
 
+impl std::convert::TryFrom<GValue> for f32 {
```
Comment (criminosis) [inline gremlin-client/src/structure/value.rs:337]: Just a heads up this is a duplicate implementation of the same code in https://github.com/wolf4ood/gremlin-rs/pull/129/files to keep the PRs independent for review and merging. One of the implementations will need to be removed.

File: gremlin-client/src/structure/value.rs [hunk 4]
```diff
@@ -333,6 +334,36 @@ impl std::convert::TryFrom<GValue> for i64 {
...
+            GValue::List(s) => from_list(s),
+            _ => Err(GremlinError::Cast(format!(
+                "Cannot cast {:?} to f32",
+                value
+            ))),
+        }
+    }
+}
+
+impl std::convert::TryFrom<GValue> for f64 {
```
Comment (criminosis) [inline gremlin-client/src/structure/value.rs:352]: Just a heads up this is a duplicate implementation of the same code in https://github.com/wolf4ood/gremlin-rs/pull/129/files to keep the PRs independent for review and merging. One of the implementations will need to be removed.

File: gremlin-client/src/conversion.rs (hunk 2 above)
Comment (wolf4ood) [reply gremlin-client/src/conversion.rs:122]: :+1:

**Comment by criminosis** [comment]
Just a heads up @wolf4ood I took an opinionated solution to how to handle mapping the lack of the property into a list/set in https://github.com/wolf4ood/gremlin-rs/pull/131/commits/8346e7111aedf5e9c55cf10f5183e4ba7ad6304d as the commit message suggests.

Another option would be to instead require those fields be wrapped in an Option so that the derive module is able to map the lack of the field to a None. Open to going that route but figured if that would be preferable we'd want to review https://github.com/wolf4ood/gremlin-rs/pull/131/commits/8346e7111aedf5e9c55cf10f5183e4ba7ad6304d and then do that after merging https://github.com/wolf4ood/gremlin-rs/pull/129 basically add support for Option<Set/List<T>> if you want to keep the case of the lack of a field with cardinality as distinct from the field having no values.

**Comment by wolf4ood** [comment]
Hi @criminosis 

threre are some conflict i think due the merge of https://github.com/wolf4ood/gremlin-rs/pull/129

Once resolved i will merge it :)

File: gremlin-client/src/structure/value.rs (hunk 4 above)
Comment (criminosis) [reply gremlin-client/src/structure/value.rs:352]: Resolved in dca681f

**Comment by criminosis** [comment]
@wolf4ood building successful again. I did a rebase and force push to keep the commit history simple and linear. Ready for your review at your leisure. Think you'll cut an 0.8 release?

With regards to the test coverage decrease, not what I can do? I'm already doing asserts for each type of the permitted data types I believe. Let me know if there's something I can improve.

**Comment by wolf4ood** [comment]
Thanks :)

**Comment by wolf4ood** [comment]
@criminosis  yes i'm releasing 0.8 :)

**Pull request merged**